Tools/Scripts/layout-test-failure-analysis

```diff
@@ -0,0 +1,291 @@
+                        required=False)
+    parser.add_argument('--stdio-export-path',
+                        dest="stdio_export_path",
+                        help='Specifies a path to which to save the raw stdio received from the server. This has no effect when loading data from a file.',
+                        required=False)
+    parser.add_argument('--combine-workers',
+                        action='store_true',
+                        help='Exports one file containing all tests in the order they were run (regardless of worker number).',
+                        required=False)
+    parser.add_argument('--no-warn',
+                        action='store_true',
+                        help='Does not warn or prompt the user before performing any actions. Potentially destructive due to bypassing file overwrite prompts. Good for automation. Will cause the script to quit if user input required.',
+                        required=False)
+
+    return parser.parse_args()
+
+
+# logging shortcuts
+def info(msg):
+    logging.info(msg)
+def warning(msg):
+    logging.warning(f'WARNING: {msg}')
+def error(msg):
+    logging.error(f'ERROR: {msg}')
+
+
+def graceful_exit(code=0):
+    print()
+    sys.exit(code)
+
+
+# Ensures a directory exists by checking and attempting to create if it doesn't exist (recursively).
+# Returns true if it exists at the end of the attempt, false if not.
+def ensure_directory_exists(dir):
+    if not os.path.isdir(dir):
+        os.makedirs(dir, exist_ok=True)
+        if not os.path.isdir(dir):
+            return False
+    return True
+
+
+# Takes a string and exports to specified file path.
+# Prompts the user for a new file path if there's an IO error.
+def export_file(path, data, no_warn = False):
+    path
+    while True:
+        try:
+            path = os.path.abspath(os.path.expanduser(path))
+            dir = os.path.dirname(path)
+            if os.path.isfile(path) and not no_warn: # file exists
+                choice = input(f'A file already exists at the location {path}. Is it okay to overwrite it? (Y/[N]): ').lower()
+                if choice.lower() != 'y' and choice.lower() != 'yes':
+                    raise ValueError()
+            if not ensure_directory_exists(dir): # directory doesn't exist
+                warning(f'Directory does not exist at {dir} and an error occurred while attempting to create it. Check permissions/spelling and try again.')
+                raise IOError()
+
+            # write file
+            print(path)
+            f = open(path, "w")
+            f.write(data)
+            f.close()
+            return path
+        except IsADirectoryError:
+            warning(f'The specified path "{path}" is a directory. Try entering a new file path.')
+        except OSError:
+            warning(f'The specified location "{path}" is read-only. Try entering a new file path.')
+        except ValueError as e:
```

Comment:
No reason to use `as e` if we aren't going to use the `e`. Also, in what circumstance would a `ValueError` be raised? In Python, `ValueError` is usually quite analagous to a compile error, it's often better to let such exception crash the script. In any case, certainly should not surpress this error.

Reply:
Good catch on `as e`, I think that was leftover from debugging.

The reason this is here is to handle a manual `ValueError` invocation earlier in the function. This would be thrown in the case that the user doesn't want to overwrite the file. There shouldn't be any other circumstance in which a ValueError is raised.

I originally programmed this such that there was a `finally` block, but python is weird with `finally` and `return`, so it would've executed the `finally` block even after `return` was called after a successful write operation. To fix this, I opted for the thrown `ValueError` approach.